homeassistant/components/valve/__init__.py

```diff
@@ -0,0 +1,289 @@
+        {
+            vol.Required(ATTR_POSITION): vol.All(
+                vol.Coerce(int), vol.Range(min=0, max=100)
+            )
+        },
+        "async_set_valve_position",
+        [ValveEntityFeature.SET_POSITION],
+    )
+
+    component.async_register_entity_service(
+        SERVICE_STOP_VALVE, {}, "async_stop_valve", [ValveEntityFeature.STOP]
+    )
+
+    component.async_register_entity_service(
+        SERVICE_TOGGLE,
+        {},
+        "async_toggle",
+        [ValveEntityFeature.OPEN | ValveEntityFeature.CLOSE],
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[ValveEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[ValveEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+@dataclass
+class ValveEntityDescription(EntityDescription):
+    """A class that describes valve entities."""
+
+    device_class: ValveDeviceClass | None = None
+
+
+class ValveEntity(Entity):
+    """Base class for valve entities."""
+
+    entity_description: ValveEntityDescription
+    _attr_current_valve_position: int | None = None
+    _attr_device_class: ValveDeviceClass | None
+    _attr_is_closed: bool | None
+    _attr_is_closing: bool | None = None
+    _attr_is_opening: bool | None = None
+    _attr_state: None = None
+    _attr_supported_features: ValveEntityFeature | None
+
+    _valve_is_last_toggle_direction_open = True
+
+    @property
+    def current_valve_position(self) -> int | None:
+        """Return current position of valve.
+
+        None is unknown, 0 is closed, 100 is fully open.
+        """
+        return self._attr_current_valve_position
+
+    @property
+    def device_class(self) -> ValveDeviceClass | None:
+        """Return the class of this entity."""
+        if hasattr(self, "_attr_device_class"):
+            return self._attr_device_class
+        if hasattr(self, "entity_description"):
+            return self.entity_description.device_class
+        return None
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the state of the valve."""
+        if self.is_opening:
+            self._valve_is_last_toggle_direction_open = True
+            return STATE_OPENING
+        if self.is_closing:
+            self._valve_is_last_toggle_direction_open = False
+            return STATE_CLOSING
+
+        if (closed := self.is_closed) is None:
+            return None
+
+        return STATE_CLOSED if closed else STATE_OPEN
+
+    @final
+    @property
+    def state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes."""
+        data = {}
+
+        if (current := self.current_valve_position) is not None:
```

Comment:
We always want to set the state attributes. If we don't have a value we set it to `None`